File: src/mirabox/integrations/google.py
"""Google Workspace 연동.

gws CLI 를 통해 읽는다. 인증은 gws 가 OAuth 키링으로 들고 있다.
"""

from __future__ import annotations

from datetime import datetime, timedelta, timezone
import json

from ..core.registry import key, source
from ..core.render import DANGER, DIM, OK, SIGNAL, blank, card, remain_text
from ..core.shell import json_after_noise, run

MAIL = "google.mail"
CALENDAR = "google.calendar"


def _iso(when: datetime) -> str:
    return when.astimezone(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


@source(MAIL, every=60)
def _fetch_mail():
    """라벨 한 번만 읽으면 되므로 가장 싸다."""
    data = json_after_noise(
        run("""gws gmail users labels get --params '{"userId":"me","id":"INBOX"}' 2>/dev/null""", 20))
    if not data:
        raise RuntimeError("gmail 응답을 파싱하지 못했다")
    return {"unread": data.get("messagesUnread", 0),
            "threads": data.get("threadsUnread", 0)}


@source(CALENDAR, every=60)
def _fetch_calendar():
    now = datetime.now(timezone.utc)
    params = json.dumps({
        "calendarId": "primary", "timeMin": _iso(now),
        "timeMax": _iso(now + timedelta(days=7)),
        "maxResults": 10, "singleEvents": True, "orderBy": "startTime",
    })
    data = json_after_noise(
        run(f"gws calendar events list --params '{params}' 2>/dev/null", 20))
    if data is None:
        raise RuntimeError("calendar 응답을 파싱하지 못했다")
    for item in data.get("items", []):
        start = (item.get("start") or {}).get("dateTime")
        if not start:
            continue                          # 종일 일정은 제외한다
        when = datetime.fromisoformat(start)
        if when > now:
            return {"in_min": round((when - now).total_seconds() / 60),
                    "title": item.get("summary", "")}
    return {"in_min": None, "title": ""}


@key("mail", "MAIL", "안 읽은 메일 수", sources=(MAIL,))
def _mail(index, state, options):
    value = state.get(MAIL)
    if not value:
        return blank(index, "MAIL")
    unread = value["unread"]
    color = OK if unread == 0 else (DANGER if unread >= 100 else SIGNAL)
    return card(index, label="MAIL", value=str(unread), value_color=color,
                right=str(value["threads"]), band_color=color)


@key("cal", "CAL", "다음 일정까지 남은 시간", sources=(CALENDAR,))
def _cal(index, state, options):
    value = state.get(CALENDAR)
    if not value:
        return blank(index, "CAL")
    if value["in_min"] is None:
        return card(index, label="CAL", value="none", value_color=DIM)
    minutes = value["in_min"]
    color = DANGER if minutes <= 15 else (SIGNAL if minutes <= 60 else OK)
    return card(index, label="CAL", value=remain_text(minutes),
                value_color=color, band_color=color)

File: src/main/integrations/google.ts
/** Google Workspace 연동. gws CLI 로 읽는다. 인증은 gws 가 OAuth 키링으로 들고 있다. */

import { ACCENT, DANGER, OK, TERTIARY, WARN, blank, card, remainText } from '../render.js'
import { key, pick, source } from '../registry.js'
import { jsonAfterNoise, sh } from '../shell.js'

export const MAIL = 'google.mail'
export const CALENDAR = 'google.calendar'

const iso = (at: number) => new Date(at).toISOString().replace(/\.\d{3}Z$/, 'Z')

interface MailValue { unread: number; threads: number }
interface CalendarValue { inMin: number | null; title: string }

source(MAIL, 60, async (): Promise<MailValue> => {
  // 라벨 한 번만 읽으면 되므로 가장 싸다
  const out = await sh(
    `gws gmail users labels get --params '{"userId":"me","id":"INBOX"}' 2>/dev/null`, 20_000)
  const data = jsonAfterNoise(out) as any
  if (!data) throw new Error('gmail 응답을 파싱하지 못했다')
  return { unread: data.messagesUnread ?? 0, threads: data.threadsUnread ?? 0 }
})

source(CALENDAR, 60, async (): Promise<CalendarValue> => {
  const now = Date.now()
  const params = JSON.stringify({
    calendarId: 'primary', timeMin: iso(now), timeMax: iso(now + 7 * 24 * 3600_000),
    maxResults: 10, singleEvents: true, orderBy: 'startTime',
  })
  const data = jsonAfterNoise(
    await sh(`gws calendar events list --params '${params}' 2>/dev/null`, 20_000)) as any
  if (!data) throw new Error('calendar 응답을 파싱하지 못했다')
  for (const item of data.items ?? []) {
    const start = item.start?.dateTime
    if (!start) continue // 종일 일정은 제외한다
    const at = Date.parse(start)
    if (at > now) return { inMin: Math.round((at - now) / 60000), title: item.summary ?? '' }
  }
  return { inMin: null, title: '' }
})

key({
  name: 'mail', label: 'MAIL', summary: '안 읽은 메일 수', sources: [MAIL],
  render: (index, state) => {
    const value = pick<MailValue>(state, MAIL)
    if (!value) return blank(index, 'MAIL')
    const color = value.unread === 0 ? OK : value.unread >= 100 ? DANGER : WARN
    return card(index, {
      label: 'MAIL', value: String(value.unread), valueColor: color,
      right: String(value.threads), bandColor: color,
    })
  },
})

key({
  name: 'cal', label: 'CAL', summary: '다음 일정까지 남은 시간', sources: [CALENDAR],
  render: (index, state) => {
    const value = pick<CalendarValue>(state, CALENDAR)
    if (!value) return blank(index, 'CAL')
    if (value.inMin == null) return card(index, { label: 'CAL', value: 'none', valueColor: TERTIARY })
    const color = value.inMin <= 15 ? DANGER : value.inMin <= 60 ? WARN : OK
    return card(index, {
      label: 'CAL', value: remainText(value.inMin), valueColor: color, bandColor: color,
    })
  },
})

void ACCENT
